File: packages/healops_opentelemetry_python/healops_opentelemetry/logger.py
import requests
import json
import threading
import time
import queue
import logging
import sys
import traceback
from datetime import datetime
from typing import Optional, Dict, Any, List

from .utils import get_caller_info, clean_stack_trace, extract_file_path_from_stack

class HealOpsLogger:
    """HealOps Logger for sending logs directly to the backend with batching support"""
    
    def __init__(self,
                 api_key: str,
                 service_name: str,
                 endpoint: str = "https://engine.healops.ai",
                 source: str = "python",
                 enable_batching: bool = True,
                 batch_size: int = 50,
                 batch_interval_ms: int = 1000,
                 environment: Optional[str] = None,
                 release: Optional[str] = None):

        self.api_key = api_key
        self.service_name = service_name
        self.endpoint = endpoint
        self.source = source
        self.environment = environment
        self.release = release

        # Batching configuration
        self.enable_batching = enable_batching
        self.batch_size = max(1, min(batch_size, 1000))
        self.batch_interval_ms = max(100, min(batch_interval_ms, 60000))

        self.log_queue = queue.Queue()
        self.is_running = False
        self._batch_thread = None

        if self.enable_batching:
            self._start_batch_processor()

    def _start_batch_processor(self):
        self.is_running = True
        self._batch_thread = threading.Thread(target=self._process_batch, daemon=True)
        self._batch_thread.start()

    def _process_batch(self):
        """Background thread to process logs in batches"""
        batch = []
        last_flush = time.time() * 1000

        while self.is_running or not self.log_queue.empty():
            try:
                # Wait for log with timeout to ensure interval flush
                timeout = max(0, (self.batch_interval_ms - (time.time() * 1000 - last_flush)) / 1000.0)
                try:
                    payload = self.log_queue.get(timeout=timeout)
                    batch.append(payload)
                except queue.Empty:
                    pass

                current_time = time.time() * 1000
                time_diff = current_time - last_flush

                # Flush if batch size reached or interval passed
                if len(batch) >= self.batch_size or (len(batch) > 0 and time_diff >= self.batch_interval_ms):
                    self._flush_batch(batch)
                    batch = []
                    last_flush = current_time

            except Exception as e:
                # Prevent thread from dying
                if os.getenv("HEALOPS_DEBUG"):
                    print(f"HealOps batch processor error: {e}")
                time.sleep(1)

    def _flush_batch(self, batch: List[Dict]):
        """Send a batch of logs to backend"""
        if not batch:
            return

        url = f"{self.endpoint}/ingest/logs/batch"
        try:
            response = requests.post(
                url,
                json={"logs": batch},
                headers={
                    "X-HealOps-Key": self.api_key,
                    "Content-Type": "application/json"
                },
                timeout=5
            )
            response.raise_for_status()
            if os.getenv("HEALOPS_DEBUG"):
                 print(f"HealOps flushed {len(batch)} logs")
        except Exception as e:
            # Fallback to individual sending if batch fails
            if os.getenv("HEALOPS_DEBUG"):
                print(f"Batch send failed, falling back to individual: {e}")
            for log in batch:
                self._send_single_log(log)

    def destroy(self):
        """Cleanup resources"""
        self.is_running = False
        if self._batch_thread:
            self._batch_thread.join(timeout=2.0)

    def info(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        """Send an INFO level log"""
        self._log("INFO", message, metadata)
    
    def warn(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        """Send a WARNING level log"""
        self._log("WARNING", message, metadata)
    
    def error(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        """Send an ERROR level log (will be persisted and may create incident)"""
        self._log("ERROR", message, metadata)
    
    def critical(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        """Send a CRITICAL level log (will be persisted and may create incident)"""
        self._log("CRITICAL", message, metadata)
    
    def _log(self, severity: str, message: str, metadata: Optional[Dict[str, Any]] = None):
        """Internal method to prepare log payload"""

        # Get caller info
        caller_info = get_caller_info()

        # Enrich metadata
        enriched_metadata = metadata or {}
        enriched_metadata.update(caller_info)

        # Add stack trace for errors if not present
        if severity in ["ERROR", "CRITICAL"]:
            if "errorStack" not in enriched_metadata:
                # Capture current stack if not provided
                stack = "".join(traceback.format_stack())
                enriched_metadata["stack"] = clean_stack_trace(stack)

        # Add code attributes for OTel compatibility
        if "filePath" in enriched_metadata:
            enriched_metadata["code.file.path"] = enriched_metadata["filePath"]
        if "line" in enriched_metadata:
            enriched_metadata["code.line.number"] = enriched_metadata["line"]
        if "functionName" in enriched_metadata:
            enriched_metadata["code.function.name"] = enriched_metadata["functionName"]

        payload = {
            "service_name": self.service_name,
            "severity": severity,
            "message": str(message),
            "source": self.source,
            "timestamp": datetime.utcnow().isoformat(),
            "metadata": enriched_metadata
        }
        
        if self.environment:
            payload["environment"] = self.environment
        if self.release:
            payload["release"] = self.release

        if self.enable_batching:
            self.log_queue.put(payload)
        else:
            self._send_single_log(payload)

    def _send_single_log(self, payload: Dict[str, Any]):
        """Send a single log immediately"""
        try:
            response = requests.post(
                f"{self.endpoint}/ingest/logs",
                json=payload,
                headers={
                    "X-HealOps-Key": self.api_key,
                    "Content-Type": "application/json"
                },
                timeout=3
            )
            response.raise_for_status()
        except Exception as e:
            if os.getenv("HEALOPS_DEBUG"):
                print(f"HealOps Logger failed to send log: {e}")

import os

class HealOpsLogHandler(logging.Handler):
    """Python logging handler that sends logs to HealOps"""

    def __init__(self, logger: HealOpsLogger):
        super().__init__()
        self.healops_logger = logger

    def emit(self, record):
        try:
            msg = self.format(record)

            # Map python log levels to HealOps severity
            level_map = {
                logging.DEBUG: "INFO",
                logging.INFO: "INFO",
                logging.WARNING: "WARNING",
                logging.ERROR: "ERROR",
                logging.CRITICAL: "CRITICAL"
            }
            severity = level_map.get(record.levelno, "INFO")

            metadata = {
                "logger_name": record.name,
                "filePath": record.pathname,
                "line": record.lineno,
                "functionName": record.funcName,
                "process": record.process,
                "thread": record.threadName
            }

            if record.exc_info:
                metadata["errorStack"] = "".join(traceback.format_exception(*record.exc_info))
                metadata["exception"] = {
                    "type": record.exc_info[0].__name__,
                    "message": str(record.exc_info[1]),
                    "stacktrace": metadata["errorStack"]
                }

            # Call internal _log directly to avoid double caller info extraction
            # We already have caller info from the record
            self.healops_logger._log(severity, msg, metadata)

        except Exception:
            self.handleError(record)

File: packages/healops_opentelemetry_go/logger.go
package healops

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"sync"
	"time"
)

// LogLevel defines the severity of a log
type LogLevel string

const (
	InfoLevel     LogLevel = "INFO"
	WarningLevel  LogLevel = "WARNING"
	ErrorLevel    LogLevel = "ERROR"
	CriticalLevel LogLevel = "CRITICAL"
)

// LoggerConfig Configuration for HealOps Logger
type LoggerConfig struct {
	APIKey          string
	ServiceName     string
	Endpoint        string
	Source          string
	Environment     string
	Release         string
	EnableBatching  bool
	BatchSize       int
	BatchInterval   time.Duration
}

// LogPayload represents a log message sent to the backend
type LogPayload struct {
	ServiceName string                 `json:"service_name"`
	Severity    LogLevel               `json:"severity"`
	Message     string                 `json:"message"`
	Source      string                 `json:"source"`
	Timestamp   string                 `json:"timestamp"`
	Environment string                 `json:"environment,omitempty"`
	Release     string                 `json:"release,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// Logger is the main struct for HealOps logging
type Logger struct {
	config    LoggerConfig
	logQueue  chan LogPayload
	done      chan struct{}
	wg        sync.WaitGroup
	isRunning bool
}

// NewLogger creates a new HealOps Logger
func NewLogger(config LoggerConfig) *Logger {
	if config.Endpoint == "" {
		config.Endpoint = "https://engine.healops.ai"
	}
	if config.Source == "" {
		config.Source = "go"
	}
	if config.BatchSize <= 0 {
		config.BatchSize = 50
	}
	if config.BatchInterval <= 0 {
		config.BatchInterval = 1 * time.Second
	}

	logger := &Logger{
		config:    config,
		logQueue:  make(chan LogPayload, config.BatchSize*2),
		done:      make(chan struct{}),
		isRunning: true,
	}

	if config.EnableBatching {
		logger.wg.Add(1)
		go logger.processBatch()
	}

	return logger
}

// Info logs an informational message
func (l *Logger) Info(message string, metadata map[string]interface{}) {
	l.log(InfoLevel, message, metadata)
}

// Warn logs a warning message
func (l *Logger) Warn(message string, metadata map[string]interface{}) {
	l.log(WarningLevel, message, metadata)
}

// Error logs an error message
func (l *Logger) Error(message string, metadata map[string]interface{}) {
	l.log(ErrorLevel, message, metadata)
}

// Critical logs a critical message
func (l *Logger) Critical(message string, metadata map[string]interface{}) {
	l.log(CriticalLevel, message, metadata)
}

// Shutdown gracefully shuts down the logger
func (l *Logger) Shutdown() {
	if !l.isRunning {
		return
	}
	l.isRunning = false
	close(l.done)
	l.wg.Wait()

	// Flush remaining logs in queue
    // Note: In a real implementation, we would want to ensure channel is drained
}

func (l *Logger) log(severity LogLevel, message string, metadata map[string]interface{}) {
	if !l.isRunning {
		return
	}

    // Enrich metadata with caller info (simplified for now)
    // In a full implementation, we would use runtime.Caller here
    if metadata == nil {
        metadata = make(map[string]interface{})
    }

	payload := LogPayload{
		ServiceName: l.config.ServiceName,
		Severity:    severity,
		Message:     message,
		Source:      l.config.Source,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
		Environment: l.config.Environment,
		Release:     l.config.Release,
		Metadata:    metadata,
	}

	if l.config.EnableBatching {
		select {
		case l.logQueue <- payload:
		default:
			// Queue full, drop log or send directly (fallback)
			if os.Getenv("HEALOPS_DEBUG") != "" {
				fmt.Println("HealOps log queue full, dropping log")
			}
		}
	} else {
		l.sendSingleLog(payload)
	}
}

func (l *Logger) processBatch() {
	defer l.wg.Done()

	batch := make([]LogPayload, 0, l.config.BatchSize)
	timer := time.NewTimer(l.config.BatchInterval)
    defer timer.Stop()

	flush := func() {
		if len(batch) > 0 {
			l.sendBatch(batch)
			batch = make([]LogPayload, 0, l.config.BatchSize)
		}
	}

	for {
		select {
		case log := <-l.logQueue:
			batch = append(batch, log)
			if len(batch) >= l.config.BatchSize {
				flush()
				// Reset timer
                if !timer.Stop() {
                    <-timer.C
                }
                timer.Reset(l.config.BatchInterval)
			}
		case <-timer.C:
			flush()
			timer.Reset(l.config.BatchInterval)
		case <-l.done:
			flush()
			return
		}
	}
}

func (l *Logger) sendBatch(logs []LogPayload) {
	url := fmt.Sprintf("%s/ingest/logs/batch", l.config.Endpoint)

    jsonData, err := json.Marshal(map[string]interface{}{"logs": logs})
	if err != nil {
		if os.Getenv("HEALOPS_DEBUG") != "" {
			fmt.Printf("Error marshalling batch logs: %v\n", err)
		}
		return
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-HealOps-Key", l.config.APIKey)

	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		if os.Getenv("HEALOPS_DEBUG") != "" {
			fmt.Printf("Error sending batch logs: %v\n", err)
		}
		// Fallback to single logs?
        for _, log := range logs {
            l.sendSingleLog(log)
        }
		return
	}
	defer resp.Body.Close()

    if os.Getenv("HEALOPS_DEBUG") != "" {
        fmt.Printf("HealOps flushed %d logs\n", len(logs))
    }
}

func (l *Logger) sendSingleLog(payload LogPayload) {
	url := fmt.Sprintf("%s/ingest/logs", l.config.Endpoint)
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-HealOps-Key", l.config.APIKey)

	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		if os.Getenv("HEALOPS_DEBUG") != "" {
			fmt.Printf("Error sending log: %v\n", err)
		}
		return
	}
	defer resp.Body.Close()
}
